### ISSUE TITLE
-Wunreachable-code-aggressive warns about code in constexpr if

### ISSUE THREAD

**Issue by pkasting** [llvm/llvm-project#57123]
https://godbolt.org/z/6xheqcvnG

This should not generate an unreachable code warning, but does.

(Non-godbolt version: Compile the following code with `--std=c++20 -Wunreachable-code-aggressive -Werror`:)

```cpp
enum E { A = 1, B = 2 };
bool foo() {
  if constexpr (A == B) {
    return false;
  } else {
    return true;
  }
}
```

**Comment by nico** [comment]
A bit smaller:

```
% cat foo.cc
bool foo() {
  if constexpr (false) {
    return false;
  } else {
    return true;
  }
}
% out/gn/bin/clang -c foo.cc -std=c++17 -Wunreachable-code-return
foo.cc:3:12: warning: 'return' will never be executed [-Wunreachable-code-return]
    return false;
           ^~~~~
1 warning generated.
```

**Comment by nico** [comment]
Looks like we generally don't suppress unreachability diags from constexpr ifs yet:

```
% cat foo.cc
void foo() {
  int a = 0;
  if constexpr (false) {
    a = 3;
  } else {
    a = 4;
  }
}
% out/gn/bin/clang -c foo.cc -std=c++17 -Wunreachable-code
foo.cc:4:9: warning: code will never be executed [-Wunreachable-code]
    a = 3;
        ^
```

**Comment by nico** [comment]
#51445 has a bunch of notes related to this warning, maybe they're useful.

@alanzhao1 I know this warning pretty well, feel free to ask me about it if you want :)

Alternatively, these diffs show the files you'll likely need to touch:
* https://reviews.llvm.org/D111568
* https://reviews.llvm.org/D112287
* https://reviews.llvm.org/D114660

**Comment by alanzhao1** [comment]
Candidate patch: https://reviews.llvm.org/D131818

**Comment by nico** [comment]
(Why is this about c++20? Repros fine in c++17 and has nothing to do with c++20 afaict (?))

**Comment by pkasting** [comment]
On the LLVM side, it isn't; it's about C++17.  On the Chromium (more accurately, Fuchsia) side, it was about C++20 because the original repro only happened using the libc++ <bit> implementation.
